client/src/_direct_session.rs

```diff
@@ -0,0 +1,454 @@
+/// public key.
+#[derive(Clone)]
+pub struct NodeEntry<IdType> {
+    pub default_id: IdType,
+    /// Vector contains all identities including default.
+    pub identities: Vec<IdType>,
+}
+
+/// Nodes to SlotIds mapping that should be used to identify forwarding calls.
+/// TODO: Consider storing here `NodeRouting`. This would save us a need to query
+///       this struct based on NodeId from `SessionLayer` when receiving packets.
+///       Only `NodeRouting` contains information necessary for decrypting packets.
+#[derive(Clone, Default)]
+pub struct AllowedForwards {
+    slots: HashMap<SlotId, NodeEntry<NodeId>>,
+    nodes: HashMap<NodeId, SlotId>,
+}
+
+impl AllowedForwards {
+    pub fn add(&mut self, node: NodeEntry<NodeId>, slot: SlotId) {
+        // Remove previous information about node.
+        // We are removing all identities and slot. This is redundant, because in most cases
+        // using default id should be enough. This protects from situations, when `NodeEntries`
+        // for different Nodes overlap. In theory it shouldn't happen, because only one Node should
+        // have access to private key, but I think it's better to ensure data consistency here.
+        self.remove_by_slot(slot);
+        self.remove(&node.default_id);
+        for id in &node.identities {
+            self.remove(id);
+        }
+
+        for id in &node.identities {
+            self.nodes.insert(*id, slot);
+        }
+        self.slots.insert(slot, node);
+    }
+
+    pub fn remove(&mut self, node_id: &NodeId) -> Option<NodeEntry<NodeId>> {
+        if let Some(slot) = self.nodes.remove(node_id) {
+            if let Some(entry) = self.slots.remove(&slot) {
+                for id in &entry.identities {
+                    self.nodes.remove(id);
+                }
+                return Some(entry);
+            }
+        }
+        None
+    }
+
+    pub fn remove_by_slot(&mut self, slot: SlotId) -> Option<NodeId> {
+        if let Some(default_id) = self.slots.get(&slot).map(|node| node.default_id) {
+            self.remove(&default_id);
+            return Some(default_id);
+        }
+        None
+    }
+
+    pub fn get_by_slot(&self, slot: SlotId) -> Option<NodeEntry<NodeId>> {
+        self.slots.get(&slot).cloned()
+    }
+
+    pub fn get_slot(&self, node_id: &NodeId) -> Option<SlotId> {
+        self.nodes.get(node_id).cloned()
+    }
+
+    pub fn get_by_id(&self, node_id: &NodeId) -> Option<NodeEntry<NodeId>> {
+        if let Some(slot) = self.nodes.get(node_id) {
+            return self.get_by_slot(*slot);
+        }
+        None
+    }
+}
+
+/// Higher level session abstraction that handles Node identification
+/// and public keys and maps it to low-level `Session`.
+/// `DirectSession` can be used to forward packets to other Nodes,
+/// than session owner.
+#[derive(Clone)]
+pub struct DirectSession {
+    /// I would prefer to use `NodeEntry<Identity>` here, but current implementation of
```

Comment:
Please don't make first-person comments. It should probably look like:
```
It would be preferred to use `NodeEntry<Identity> here, but current implementation [...]
```

My statement applies to all PR as a whole, not only in this place.